### ISSUE TITLE
Training error on DynamicViT

### ISSUE THREAD

**Issue by King4819** [JoakimHaurum/TokenReduction#2]
Hi, I want to ask that when training on DynamicViT at keep ratio **0.3**, the training will terminate due to loss nan problem, my training script is shown below. 
Since I remember the original DynamicViT won't have this issue at keep ratio **0.3**, but I do encountered this issue on other ViT token-pruning methods. Hopes to get your reply, thanks !

<img width="494" alt="image" src="https://github.com/JoakimHaurum/TokenReduction/assets/120317730/d52d6f0f-2c4a-4e69-abfe-f64b49961a83">

CUDA_VISIBLE_DEVICES=0 \
python train.py \
    --dataset imagenet \
    --data my dataset path ... \
    --batch-size 256 \
    --lr 0.001 \
    --epochs 30 \
    --warmup-epochs 5 \
    --lr_batch_normalizer 1024 \
    --sched_in_steps \
    --use_amp \
    --grad_accum_steps 4 \
    --output_dir output \
    --model dyvit_tiny_patch16_224 \
    --reduction_loc 3 6 9 \
    --keep_rate 0.3

**Comment by JoakimHaurum** [comment]
Hi!

DynamicViT can at times be hard to train, especially with aggressive keep rates. I encountered this my self and through personal correspondence with the original authors had this confirmed. My suggestion would be to try and run with a new seed (using --seed) and with by constraining the gradient norm (using --clip-grad).

Hope this helps!

**Comment by King4819** [comment]
Thanks for your reply. I would like to ask another question. Seems that your code utilizing gradient accumulation technique, which is very helpful for those doesn't have enough computing resource. My question is: have your verified the performance of utilizing gradient accumulation ? For example, 
the training result of: --batch_size 256 --grad_accum_steps = 4 in single GPU training will be the same as --batch_size 512 in 2 GPU training ? Since deit requires total batch size of 1024 for training.
Thanks you very much and this is an awesome work !!!

**Comment by JoakimHaurum** [comment]
The gradient accumulation should be verified, as I built on the well established repos from timm and DeiT.

**Comment by benbergner** [comment]
I encountered the same issue already in the forward pass, indicating that setting --clip-grad might not be the problem.
I then noticed that the nans come from PreditorLG when the policy is all 0s and that there is a bug when adding eps to avoid division by 0.

You might consider computing global_x as follows:

`global_x = (x[:, :, C // 2 :] * policy).sum(dim=1, keepdim=True) / (torch.sum(policy, dim=1, keepdim=True) + self.eps)` 

I have not extensively tested it but nans are avoided in the first few iterations even with aggressive pruning rates.

**Comment by King4819** [comment]
@benbergner Thanks for your reply. I have observed that this problem may caused by the prediction module. If **the prediction module predicts zero token to be preserved**, it will cause loss nan. So my solution is to ensure that the prediction module at least preserve one token.

**Comment by benbergner** [comment]
How exactly do you ensure this?

**Comment by King4819** [comment]
@benbergner I set a condition to check if the prediction module predicts zero token to be preserved, manually modify the decision mask to at least preserve one token.
